crates/turbopack/src/module_options/module_rule.rs

```diff
@@ -73,54 +67,53 @@ impl ModuleRuleCondition {
     }
 }
 
-#[turbo_tasks::value_impl]
 impl ModuleRuleConditionVc {
-    #[turbo_tasks::function]
-    pub async fn matches(self, path: FileSystemPathVc) -> Result<BoolVc> {
+    pub async fn matches(self, path: FileSystemPathVc) -> Result<bool> {
         let path_ref = path.await?;
         Ok(match &*self.await? {
-            ModuleRuleCondition::All(conditions) => BoolVc::cell(
-                conditions
-                    .iter()
-                    .map(|c| c.matches(path))
-                    .try_join()
-                    .await?
-                    .into_iter()
-                    .all(|c| *c),
-            ),
-            ModuleRuleCondition::Any(conditions) => BoolVc::cell(
-                conditions
-                    .iter()
-                    .map(|c| c.matches(path))
-                    .try_join()
-                    .await?
-                    .into_iter()
-                    .any(|c| *c),
-            ),
-            ModuleRuleCondition::ResourcePathEndsWith(end) => {
-                BoolVc::cell(path_ref.path.ends_with(end))
-            }
+            ModuleRuleCondition::All(conditions) => conditions
+                .iter()
+                .map(|c| c.matches_boxed(path))
+                .try_join()
+                .await?
+                .into_iter()
+                .all(|c| c),
+            ModuleRuleCondition::Any(conditions) => conditions
+                .iter()
+                .map(|c| c.matches_boxed(path))
+                .try_join()
+                .await?
+                .into_iter()
+                .any(|c| c),
+            ModuleRuleCondition::ResourcePathEndsWith(end) => path_ref.path.ends_with(end),
             ModuleRuleCondition::ResourcePathHasNoExtension => {
-                BoolVc::cell(if let Some(i) = path_ref.path.rfind('.') {
+                if let Some(i) = path_ref.path.rfind('.') {
                     if let Some(j) = path_ref.path.rfind('/') {
                         j > i
                     } else {
                         false
                     }
                 } else {
                     true
-                })
+                }
             }
-            ModuleRuleCondition::ResourcePathInDirectory(dir) => BoolVc::cell(
+            ModuleRuleCondition::ResourcePathInDirectory(dir) => {
                 path_ref.path.starts_with(&format!("{dir}/"))
-                    || path_ref.path.contains(&format!("/{dir}/")),
-            ),
+                    || path_ref.path.contains(&format!("/{dir}/"))
+            }
             ModuleRuleCondition::ResourcePathInExactDirectory(parent_path) => {
-                path.is_inside(*parent_path)
+                *path.is_inside(*parent_path).await?
             }
             _ => todo!("not implemented yet"),
         })
     }
+
+    pub fn matches_boxed(
```

Comment:
I'm adding the [async_recursion]() crate in https://github.com/vercel/turbo/pull/2426, because this [pinboxdynfut](https://docs.rs/async-recursion/latest/async_recursion/) hurts to think about.

Reply:
I'm getting rid of all `async` in this module next, as we don't need to store `Vc`s. We can store `ReadRef`s instead.